dali/pipeline/workspace/workspace.h

```diff
@@ -145,12 +145,38 @@ class WorkspaceBase : public ArgumentWorkspace {
   /**
    * @brief Returns the number of inputs.
    */
-  inline int NumInput() const { return input_index_map_.size(); }
+  inline int NumInput() const {
+    return input_index_map_.size();
+  }
 
   /**
    * @brief Returns the number of outputs.
    */
-  inline int NumOutput() const { return output_index_map_.size(); }
+  inline int NumOutput() const {
+    return output_index_map_.size();
+  }
+
+  /**
+   * Returns shape of input at given index
+   * @return TensorShape<> for SampleWorkspace, TensorListShape<> for other Workspaces
+   */
+  template <typename Backend>
+  auto GetInputShape(int input_idx) const {
+    const auto& input = *InputHandle(input_idx, Backend{});
+    return input.shape();
+  }
+
+  /**
+   * Returns batch size for given input
+   */
+  template <typename Backend>
+  int get_batch_size(int input_idx) const {
+    return NumInput() == 0 ? batch_size_ : GetInputShape<Backend>(input_idx).num_samples();
+  }
```

Comment:
I find this function disturbing:
1. It returns something else than the argument says (when there are no inputs it returns something else) - perhaps there should be a parameterless overload.
2. It returns _input_ batch size, not batch size - and the two will differ quite soon (conditional ops?). Let's plan a few steps ahead and not make ourselves suffer lengthy rebuilds, API changes and conflicts. Now the function is not used yet by any upstream code and it's the right time to make the API future-proof.
3. Instead of manually providing backend, we can check which backend we have - this will save us a lot of template disambiguations: `ws.template get_batch_size<Backend>()` -> `ws.get_batch_size()`

Let's discuss it tomorrow.